crates/sui-graphql-rpc/src/types/dot_move/named_type.rs

```diff
@@ -0,0 +1,203 @@
+
+        // now let's create a hashmap with {name: MovePackage}
+        let mut name_package_id_mapping = HashMap::new();
+
+        // doing it in reverse so we can pop instead of shift
+        for name in names.into_iter().rev() {
+            // safe unwrap: we know that the amount of results has to equal the amount of names.
+            let Some(package) = results.pop().unwrap() else {
+                return Err(Error::DotMove(DotMoveServiceError::NameNotFound(name)));
+            };
+
+            name_package_id_mapping.insert(name, package.native.id());
+        }
+
+        let correct_type_tag = Self::replace_names(name, &name_package_id_mapping);
+
+        // now we query the names with futures to utilize data loader
+        Ok(correct_type_tag)
+    }
+
+    // TODO: Should we introduce some overall string limit length here?
+    // Is this already caught by the global limits?
+    // This parser just extracts all names from a type tag, and returns them
+    // We do not care about de-duplication, as the dataloader will do this for us.
+    // The goal of replacing all of them with `0x0` is to make sure that the type tag is valid
+    // so when replaced with the move name package addresses, it'll also be valid.
+    fn parse_names(name: &str) -> Result<Vec<String>, Error> {
+        let mut names = vec![];
+        let struct_tag = VERSIONED_NAME_UNBOUND_REG.replace_all(name, |m: &regex::Captures| {
+            // safe unwrap: we know that the regex will always have a match on position 0.
+            names.push(m.get(0).unwrap().as_str().to_string());
+            "0x0".to_string()
+        });
+
+        // We attempt to parse the type_tag with these replacements, to make sure there are no other
+        // errors in the type tag (apart from the move names). That protects us from unnecessary
+        // queries to resolve .move names, for a type tag that will be invalid anyway.
+        parse_type_tag(&struct_tag).map_err(|e| Error::Client(e.to_string()))?;
+
+        Ok(names)
+    }
+
+    // This function replaces all the names in the type tag with their corresponding MovePackage address.
+    // The names are guaranteed to be the same and exist (as long as this is called in sequence),
+    // since we use the same parser to extract the names.
+    fn replace_names(type_name: &str, names: &HashMap<String, ObjectID>) -> String {
+        let struct_tag_str =
+            VERSIONED_NAME_UNBOUND_REG.replace_all(type_name, |m: &regex::Captures| {
+                // safe unwrap: we know that the regex will have a match on position 0.
+                let name = m.get(0).unwrap().as_str();
+
+                // if we are miss-using the function, and we cannot find the name in the hashmap,
+                // we return an empty string, which will make the type tag invalid.
+                if let Some(addr) = names.get(name) {
+                    addr.to_string()
+                } else {
+                    "".to_string()
```

Comment:
Let's make this an explicit error please -- otherwise the error we output to users is going to be confusing.

Reply:
I think the problem I had here was that the internal function on `regexp.replace_all` doesn't use  "Result", so I wasn't sure how to throw an error out? Any suggestions?

Reply:
Did some research and found this: https://github.com/rust-lang/regex/issues/648#issuecomment-590072186

It seems to be a problem that people quite regularly bump up against, and the solution seems to be to write your own replacement routine -- it doesn't seem too too bad to do that. The other option I can think of is to go over the captures in two passes -- once, in a loop to fetch all the names in order, and store them in a `Vec` and then next to do the replacement, and you can pop the names off and use them one at a time.